### PR TITLE
Extend GeoPolygon to support interiors

### PR DESCRIPTION
**Pull request by zzzz-vincent**
We should support polygon with exterior and interiors (holes on the surface) in Geo Filter by Polygon(https://github.com/qdrant/qdrant/issues/795). This commit extend current GeoPolygon filter to accept interiors. It includes:
1. changes to proto and internal GeoPolygon struct, and validation fn
2. add and refactor some tests
3. add integration test

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
4. [x] Have you formatted your code locally using `cargo fmt` command prior to submission?
5. [x] Have you checked your code using `cargo clippy` command?

### Changes to Core Features:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your core changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: docs/redoc/master/openapi.json
Comment (agourlay) [inline docs/redoc/master/openapi.json:5214]: With this design we are not respecting strictly the [GeoJSON API](https://datatracker.ietf.org/doc/html/rfc7946#section-3.1.6).

```
o  For type "Polygon", the "coordinates" member MUST be an array of
      linear ring coordinate arrays.

   o  For Polygons with more than one of these rings, the first MUST be
      the exterior ring, and any others MUST be interior rings.  The
      exterior ring bounds the surface, and the interior rings (if
      present) bound holes within the surface.
````

I like your proposal better as it is more explicit.

File: lib/api/src/grpc/conversions.rs
Comment (agourlay) [inline lib/api/src/grpc/conversions.rs:921]: But the gRPC follows the standard strictly regarding exterior rings.

I wonder if we should have both REST & gRPC do the same thing.

File: lib/segment/src/index/field_index/geo_hash.rs
Comment (agourlay) [inline lib/segment/src/index/field_index/geo_hash.rs:21]: thanks got migrating away from the deprecated API :+1:

File: lib/segment/src/index/field_index/geo_hash.rs
Comment (agourlay) [inline lib/segment/src/index/field_index/geo_hash.rs:715]: I liked those comment explaining the test cases.
Would be great to keep them!

File: docs/redoc/master/openapi.json
Comment (zzzz-vincent) [reply docs/redoc/master/openapi.json:5214]: Got you! Will change to rings and have it also align with gRPC.

File: lib/api/src/grpc/conversions.rs
Comment (zzzz-vincent) [reply lib/api/src/grpc/conversions.rs:911]: Changed RESTand gRPC to the same.

File: lib/segment/src/index/field_index/geo_hash.rs
Comment (zzzz-vincent) [reply lib/segment/src/index/field_index/geo_hash.rs:715]: added back.

File: docs/redoc/master/openapi.json
Comment (zzzz-vincent) [reply docs/redoc/master/openapi.json]: I have aligned both gRPC and REST to the GeoJSON API. Currently, I've named the field in the Polygon as rings, which is directly derived from the GeoJSON specification.

**Comment by agourlay** [comment]
@zzzz-vincent It looks like you just need the integration tests for the REST API at this point.

I wonder if you should just add those tests in this PR as it is otherwise not tested at this level.

Having more confidence that the API behaves properly would help the review I believe.
The drawback is that it would this PR bigger.

WDYT about this idea?

**Comment by agourlay** [comment]
FYI the polygon API changes which landed in dev are being commented out to smooth the next minor release.
https://github.com/qdrant/qdrant/pull/2352

File: lib/collection/src/operations/validation.rs
Comment (agourlay) [inline lib/collection/src/operations/validation.rs:92]:
```suggestion
        "min_ring_length" => match (params.get("min_length"), params.get("length"), params.get("index")) {
```

File: openapi/tests/openapi_integration/test_geo_filter.py
Comment (agourlay) [inline openapi/tests/openapi_integration/test_geo_filter.py:18]: I'd love a few examples with malformed polygons to check they are detected and reported properly

File: lib/segment/src/types.rs
Comment (coszio) [inline lib/segment/src/types.rs:1239]: Why not create two fields?

File: lib/segment/src/payload_storage/condition_checker.rs
Comment (coszio) [inline lib/segment/src/payload_storage/condition_checker.rs:172]: Since this is the third time this code is repeated, we can abstract it away. I will make a suggestion on a separate PR

File: openapi/tests/openapi_integration/helpers/collection_setup.py
Comment (coszio) [inline openapi/tests/openapi_integration/helpers/collection_setup.py:97]: maybe the repo should specify a python formatter, cc @generall

**Review by coszio — COMMENTED** [review]
Sorry for the delayed review, see comments

**Comment by coszio** [comment]
^ [suggestions PR](https://github.com/zzzz-vincent/qdrant/pull/1), as promised

File: lib/segment/src/payload_storage/condition_checker.rs
Comment (coszio) [reply lib/segment/src/payload_storage/condition_checker.rs:172]: https://github.com/zzzz-vincent/qdrant/pull/1, as promised

File: lib/segment/src/types.rs
Comment (zzzz-vincent) [reply lib/segment/src/types.rs:1239]: The [GeoJSON spec](https://datatracker.ietf.org/doc/html/rfc7946#section-3.1.6) defines as first ring and other rings. So I think have gRPC proto same as GeoJSON can be more cleaner.

File: openapi/tests/openapi_integration/test_geo_filter.py
Comment (zzzz-vincent) [reply openapi/tests/openapi_integration/test_geo_filter.py:18]: added.

File: lib/segment/src/payload_storage/condition_checker.rs
Comment (zzzz-vincent) [reply lib/segment/src/payload_storage/condition_checker.rs:172]: ack. Will do in the next PR.

File: openapi/tests/openapi_integration/helpers/collection_setup.py
Comment (zzzz-vincent) [reply openapi/tests/openapi_integration/helpers/collection_setup.py:97]: sounds good. I added an issue for this https://github.com/qdrant/qdrant/issues/2451

File: lib/segment/src/types.rs
Comment (coszio) [reply lib/segment/src/types.rs:1293]: Oh, didn't know about this, might be worth to explicitly say this is following GeoJSON spec.

Comment (zzzz-vincent) [reply lib/segment/src/types.rs:1294]: Yes this makes sense! added.

File: lib/segment/src/types.rs
Comment (agourlay) [inline lib/segment/src/types.rs:1316]: why do we need this `GeoPolygonShadow`?

**Comment by agourlay** [comment]
We have sorted out the discussion around GeoJSON with @zzzz-vincent on Discord.

We have been oscillating for a while between adhering strictly to the GeoJSON spec and respecting the existing API consistency on this PR. 

This has caused some unfortunate back-and-forth and confusion in our design.

**Here is the bottom line, we actually want to prioritize the consistency with our existing Geo API over respecting the GeoJSON spec.**

It is more important that the API makes to users than to achieve spec. compliance.

First, given the pre-existence of `geo_bounding_box` and `geo_radius` in our API, we cannot introduce a new polymorphic type covering the corresponding [GeoJSON geometries](https://datatracker.ietf.org/doc/html/rfc7946#section-3.1).

Those variants require a `coordinates` field and use a different encoding with `lat` and `lon` fields.

Implementing only a subset of the geometries in a different way is highly confusing.

Secondly, encoding polygons using an array of arrays for the rings definition creates an error prone interface as it relies on the ordering of the element to determine which ring is outer and which rings are inner.

We prefer having clear dedicated fields `exterior` and `interiors` instead.

Third, the UX of `geo_polygon` queries should feel like an extension of the `geo_bounding_box` and `geo_radius`.

There is no special case input type for `multi_geo_bounding_boxes` or `multi_geo_radiuses`.

In order to query several geometries, one must use several match filters.

Therefore, `multi_polygons` should be handled the same way and not have a dedicated input type.

**The conclusion is that we will not follow exactly the GeoJSON spec and prioritize consistency and ease of use.**

Kudos to @zzzz-vincent for providing great arguments.

File: lib/segment/src/types.rs
Comment (agourlay) [reply lib/segment/src/types.rs:1325]: follow up in https://github.com/qdrant/qdrant/pull/2315#issuecomment-1689553067

File: lib/segment/src/types.rs
Comment (zzzz-vincent) [reply lib/segment/src/types.rs:1317]: I am following this pr: https://github.com/qdrant/qdrant/pull/295. Not sure if there are other better ways but directly implement new() and do validation inside doesn't really work.

File: lib/segment/src/types.rs
Comment (zzzz-vincent) [reply lib/segment/src/types.rs]: done.

File: lib/api/src/grpc/validate.rs
Comment (zzzz-vincent) [reply lib/api/src/grpc/validate.rs]: ack :o

File: lib/segment/src/types.rs
Comment (agourlay) [inline lib/segment/src/types.rs:1310]: What about making `interiors` optional?
It seems to me most people would not use it, it would make the API lighter not having to provide an empty list.

WDYT?

File: openapi/tests/openapi_integration/test_geo_filter.py
Comment (agourlay) [inline openapi/tests/openapi_integration/test_geo_filter.py:189]: maybe you could copy paste this query with `must` to show it returns an empty set.

Comment (zzzz-vincent) [reply openapi/tests/openapi_integration/test_geo_filter.py:224]: added.

File: lib/segment/src/types.rs
Comment (zzzz-vincent) [reply lib/segment/src/types.rs]: yes, I think this is a good idea! Changed to Option and also modify integration test to ensure skipping interiors is okay.

**Comment by agourlay** [comment]
@zzzz-vincent This looks great , thank you very much for this great effort :+1: 

Before merging this and updating the API on dev, I think you could check the following:
- Python client generation to make sure the current API design works there
- basic congruence tests to make sure everything is fine

**Review by agourlay — APPROVED** [review]
I have validated that the python client can be generated :+1:

File: lib/api/src/grpc/proto/points.proto
Comment (coszio) [inline lib/api/src/grpc/proto/points.proto:582]:
```suggestion
  GeoLineString exterior = 1; // The exterior ring bounds the surface
```

**Review by coszio — CHANGES_REQUESTED** [review]
Only comments about the docs and tests, otherwise it's great work :raised_hands:

File: lib/segment/src/types.rs
Comment (IvanPleshkov) [inline lib/segment/src/types.rs:1324]: Instead of custom `GeoPolygonValidationError` I propose to use `OperationError` defined in `segment`crate. If no good variant exists in `OperationError` enum, you can add a new variant. `OperationError` has all conventions and unique formatting format. And instead of `Result<_, OperationError>` you can use `OperationResult<_>`. What do you think about this error handling @agourlay ?

File: lib/segment/src/types.rs
Comment (IvanPleshkov) [inline lib/segment/src/types.rs:1347]: I see that the first and last points should be equal. And I see that `line.points.len()` can be 3. But if `line.points.len() == 3` and `line.points.first() != line.points.last()`, user provided a single line instead of triangle. I guess `line.points.len() == 3` is incorrect check and it should be `>=4`

Comment (IvanPleshkov) [reply lib/segment/src/types.rs:1347]: Also, should we have float equality here without any epsilon?

**Review by IvanPleshkov — CHANGES_REQUESTED** [review]

File: lib/segment/src/types.rs
Comment (zzzz-vincent) [reply lib/segment/src/types.rs]: changed to epsilon

File: openapi/tests/openapi_integration/test_geo_filter.py
Comment (zzzz-vincent) [reply openapi/tests/openapi_integration/test_geo_filter.py:56]: splitted to several tests

File: lib/api/src/grpc/proto/points.proto
Comment (zzzz-vincent) [reply lib/api/src/grpc/proto/points.proto]: 1) Yes,  interiors can partially overlap the exterior.
2) updated comments.

**Review by coszio — APPROVED** [review]

**Review by timvisee — APPROVED** [review]
Nice work!

Note that I only skimmed the code. I mostly looked at the API design.

**Review by IvanPleshkov — APPROVED** [review]

File: lib/segment/src/types.rs
Comment (agourlay) [reply lib/segment/src/types.rs:1331]: I think Ivan makes a good point :+1: 

We already have an existing infrastructure for error, it would be nice to use it instead of this ad-hoc representation.

@zzzz-vincent this should be the very last change!

Comment (zzzz-vincent) [reply lib/segment/src/types.rs:1331]: Sure, have reflected this in current change. There are also some other struct using customized Error such as `PointsSelectorValidationError` and `GeoPointValidationError`. We should have a seperate PR to refactor them.

**Comment by agourlay** [comment]
@zzzz-vincent We have an implementation hole as described in https://github.com/qdrant/qdrant/issues/2539#issuecomment-1700689177

IMHO we can't merge this PR without it.

You could add a first implementation relying only the outer ring to always have a cardinality.

**Comment by generall** [comment]
Hi @zzzz-vincent!
I agree with @agourlay, the cardinality estimation is required. Important point that @agourlay explained to me, is that internal polygons are always within the external one, that should make the cardinality estimation a bit simpler and more accurate than exclusion with `must_not`. But without proper carnality estimation query planner won't be able to choose efficient strategy for executing the query and the accuracy can easily drop to 0.

In addition to this, I would suggest to optimize check_condition function here: https://github.com/qdrant/qdrant/blob/809b04c062bfa80b0b97bc01b417f4d2c5c70bb4/lib/segment/src/index/field_index/geo_index.rs#L563

I noticed, that it builds a Polygon object on each call, which could be wasteful, considering that this maybe be called thousand+ times during the single search. I would recommend to either pre-generate polygon inside `filter`, or keep Polygon as an additional optional field of `GeoPolygon`, but exclude it from (de-)serialization.

**Comment by zzzz-vincent** [comment]
Thanks for pointing out @generall. A new change include cardinality and unit tests has pused.
For the duplication of building Polygon, I added a wrapper class of `PolygonWrapper` and make `GeoPolygon` can convert to this class to call `check_point`.  And then it can be pre-convert from GeoPolygon before iterating all points.

File: lib/segment/src/index/field_index/geo_index.rs
Comment (generall) [inline lib/segment/src/index/field_index/geo_index.rs:563]:
```suggestion
            let geo_condition_copy = geo_polygon.convert();
```

File: lib/segment/src/index/query_optimization/condition_converter.rs
Comment (generall) [inline lib/segment/src/index/query_optimization/condition_converter.rs:179]: `polygon_wrapper` should be created outside of the lambda function, so that it will only be created once per query.

File: lib/segment/src/index/field_index/geo_index.rs
Comment (generall) [inline lib/segment/src/index/field_index/geo_index.rs:604]: Since geo-polygon has internal exclusion, it should account it in the estimation.

**Review by generall — CHANGES_REQUESTED** [review]

File: lib/segment/src/index/query_optimization/condition_converter.rs
Comment (zzzz-vincent) [reply lib/segment/src/index/query_optimization/condition_converter.rs]: done.

File: lib/segment/src/index/field_index/geo_index.rs
Comment (zzzz-vincent) [reply lib/segment/src/index/field_index/geo_index.rs]: done.

File: lib/segment/src/index/field_index/geo_index.rs
Comment (zzzz-vincent) [reply lib/segment/src/index/field_index/geo_index.rs:604]: polygon_hashes has taken internal exclusion (interiors) in to consideration. Added an additional step to convert a polygon_wrapper and use it for filtering to reduce the unnecessary building of new Polygon everytime in `check_polygon_intersection` fn.

**Comment by zzzz-vincent** [comment]
@timvisee Hi Tim, I saw you PR #2548. This is much better! The previous implementation was used just to provide a workaround for REST API as the validation is only for grpc. But you change which introduced generic validation should work!

**Comment by agourlay** [comment]
@zzzz-vincent Please rebase only if necessary during the review.
Not having a clear diff makes it difficult to see what changed :)

File: lib/segment/src/index/field_index/geo_index.rs
Comment (zzzz-vincent) [reply lib/segment/src/index/field_index/geo_index.rs]: Thanks for pointing out. I didn't realize the precision incompatibility. I pushed a version per your suggestion. However, I am not very confident about "subtract interior estimation". Currently, all tests are passed. But for the new test `test_polygon_with_exclusion`, if I increase the value of `GEO_QUERY_MAX_REGION`, the `real_cardinality` may lower than card.min.

Comment (agourlay) [reply lib/segment/src/index/field_index/geo_index.rs]: > But for the new test test_polygon_with_exclusion, if I increase the value of GEO_QUERY_MAX_REGION, the real_cardinality may lower than card.min.

@zzzz-vincent I have not been able to reproduce this case, which `GEO_QUERY_MAX_REGION` did you use?

File: lib/segment/src/index/field_index/geo_index.rs
Comment (agourlay) [inline lib/segment/src/index/field_index/geo_index.rs:613]: I am always confused my computation of the expected cardinality.

Why do you `exterior_estimation.min` here?

File: lib/segment/src/index/field_index/geo_hash.rs
Comment (agourlay) [inline lib/segment/src/index/field_index/geo_hash.rs:267]: This is a pattern prevalent in this file, as discussed we should remove them all at once in a follow up PR.

Tracked here https://github.com/qdrant/qdrant/issues/2645

Comment (zzzz-vincent) [reply lib/segment/src/index/field_index/geo_hash.rs:267]: ack.

File: lib/segment/src/index/field_index/geo_index.rs
Comment (zzzz-vincent) [reply lib/segment/src/index/field_index/geo_index.rs:613]: since we also needs to ensure the exp is in the range of [min, max]. This logic also apply to `match_cardinality` https://github.com/qdrant/qdrant/blob/a6164ccda89b00eb361686ed40b90da9fbb1b2c3/lib/segment/src/index/field_index/geo_index.rs#L309

File: lib/segment/src/index/field_index/geo_index.rs
Comment (zzzz-vincent) [reply lib/segment/src/index/field_index/geo_index.rs]: Per offline discussion with @agourlay , I found that I used some very huge `GEO_QUERY_MAX_REGION` value (eg, 500). This may not be a reasonable number used for any real case.

File: lib/segment/src/index/field_index/geo_hash.rs
Comment (agourlay) [reply lib/segment/src/index/field_index/geo_hash.rs:267]: The error handling issue has been started by an external contributor, you could give it a short review :)
https://github.com/qdrant/qdrant/pull/2701

File: lib/api/src/grpc/validate.rs
Comment (agourlay) [inline lib/api/src/grpc/validate.rs:180]: I think there was a bad merge, I do not think you introduced this.

Comment (agourlay) [reply lib/api/src/grpc/validate.rs:180]: This function now lives in `lib/common/validation`

File: lib/segment/src/index/field_index/geo_hash.rs
Comment (agourlay) [reply lib/segment/src/index/field_index/geo_hash.rs:267]: The fix was merged, you can rebase and use the new error handling :)

File: lib/api/src/grpc/validate.rs
Comment (zzzz-vincent) [reply lib/api/src/grpc/validate.rs]: removed.

File: lib/segment/src/index/field_index/geo_hash.rs
Comment (zzzz-vincent) [reply lib/segment/src/index/field_index/geo_hash.rs]: merged :)

**Comment by agourlay** [comment]
@zzzz-vincent I took the liberty to fix https://github.com/qdrant/qdrant/pull/2745#discussion_r1341443464 following the merge.

**Review by generall — APPROVED** [review]

**Pull request merged**
